### PR TITLE
unique attributes

### PR DESCRIPTION
**Pull request by martin-kuba**
This adds a new general feature of attributes - attributes with unique values. The feature is used in modules for attributes member:o365EmailAddresses:mu and group_resource:o365EmailAddresses:mu.
The feature is implemented using special database tables. For each of the 12 PerunBeans or their tuples that can have attributes, there is a new table **<bean>_attr_u_values** referencing table **<bean>_attr_values**, and the new table has a unique constraint on columns with attribute definition id and value. The feature is implemented for all possible attribute types, i.e. String, Integer, Boolean, ArrayList, LinkedHashMap and their large-value variants. For ArrayList, items in list are inserted separately. For LinkedHashMap, key-value pairs are inserted separately as key=value strings.

- added unique flag to AttributeDefinition and corresponding is_unique column to table attr_names
- defined tables for unique attr values with suffix _attr_u_values for Oracle, Postgres and HSQLDB in both change logs and complete db creations
- fixed some existing attribute tables, replaced unique constraints with primary keys that can be referenced in foreign keys (user_attr_values, member_resource_attr_values, member_group_attr_values, user_facility_attr_values)
- enhanced existing tables definitions by moving table constraints to table definitions from separate ALTER TABLE ADD CONSTRAINT statements 
- wrote integration tests
- implemented handling of deletion and updates of values of unique attributes
- added API call for converting attribute to unique
- added AttributeAlreadyMarkedUniqueException
- changed WrongAttributeValueException to have message always at the end of string (needed for unit tests)
- added methods to Attribute for casting its value as all possible types
- in Logback test config removed thread and logger ids to make line shorter
- in AttributesManagerBl, added method getPerunBeanIdsForUniqueAttributeValue()  for finding ids of entities that have a given attribute value, its needed to check whether the unique value to be inserted is on the same object or some another
- added getLockTable() method in Compatibility
- implemented checking of uniqueness of values in two attributes simultaneously in the modules for o365EmailAddresses

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_resource_attribute_def_def_o365EmailAddresses_mu.java [hunk 1]
```diff
@@ -43,19 +46,22 @@
...
 
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_group_resource_attribute_def_def_o365EmailAddresses_mu.class);
+
 	private static final String NAMESPACE = AttributesManager.NS_GROUP_RESOURCE_ATTR_DEF;
 	static final String ADNAME_ATTRIBUTE = NAMESPACE + ":adName";
 
 	@Override
 	public void checkAttributeValue(PerunSessionImpl sess, Resource resource, Group group, Attribute attribute) throws InternalErrorException, WrongAttributeValueException, WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
-		List<String> emails;
+		log.trace("checkAttributeValue(resource={},group={},attribute={})",resource,group,attribute);
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_resource_attribute_def_def_o365EmailAddresses_mu.java]: This was probably only for testing purpose, do you need it in production code too?

File: perun-db/oracle.sql [hunk 2]
```diff
@@ -19,7 +19,9 @@ create table vos (
 	modified_by nvarchar2(1300) default user not null,
 	status char(1) default '0' not null,
 	created_by_uid integer,
-	modified_by_uid integer
+	modified_by_uid integer,
+	constraint VO_PK primary key (id),
+	constraint VO_U unique (name)
```
Comment (stavamichal) [inline perun-db/oracle.sql]: So now some tables has constraints in definition of table itself, and same still have constraints below? This is very confusing.

File: perun-db/oracle.sql [hunk 3]
```diff
@@ -1822,8 +1817,124 @@ alter table members_sponsored add (
...
+
+CREATE TABLE vo_attr_u_values (
+	vo_id INT NOT NULL,
+	attr_id INT NOT NULL,
+	attr_value VARCHAR(4000),
+	UNIQUE (attr_id, attr_value),
+	FOREIGN KEY (vo_id,attr_id) REFERENCES vo_attr_values ON DELETE CASCADE
+);
+CREATE INDEX vauv_idx ON vo_attr_u_values (vo_id, attr_id) ;
+
```
Comment (stavamichal) [inline perun-db/oracle.sql]: You should preserve existing order in the file or change it whole, not just part of it.

File: perun-db/postgres.sql [hunk 4]
```diff
@@ -1772,6 +1757,125 @@ alter table user_ext_source_attr_values add constraint uesattrval_attr_fk foreig
...
+CREATE INDEX ufauv_idx ON user_facility_attr_u_values (user_id, facility_id, attr_id) ;
+
+CREATE TABLE vo_attr_u_values (
+	vo_id INT NOT NULL,
+	attr_id INT NOT NULL,
+	attr_value VARCHAR(4000),
+	UNIQUE (attr_id, attr_value),
+	FOREIGN KEY (vo_id,attr_id) REFERENCES vo_attr_values ON DELETE CASCADE
+);
+CREATE INDEX vauv_idx ON vo_attr_u_values (vo_id, attr_id) ;
```
Comment (stavamichal) [inline perun-db/postgres.sql]: You should preserve existing order in the file or change it whole, not just part of it.

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Attribute.java [hunk 5]
```diff
@@ -58,6 +58,29 @@ public Object getValue() {
 		return value;
 	}
 
+	public String valueAsString() {
```
Comment (zlamalp) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/Attribute.java]: I believe, all of these methods will be automatically called, when RPC serialize object. Not sure, if it will fail or simply duplicate the value in returned JSON object, but if you explicitly need these methods, please add them to annotation ``JsonIgnoreProperties`` for ``AttributeMixin`` (project perun-base, package cz.metacentrum.perun.rpc.serializer).

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/AttributeDefinition.java [hunk 6]
```diff
@@ -38,6 +38,12 @@
 	 */
 	private boolean writable;
 
+	/**
+	 * If the attribute values must be unique. For multivalued types like java.util.ArrayList, each value in the list
+	 * for a given object must be unique among all values for all objects.
+	 */
+	private boolean unique;
```
Comment (zlamalp) [inline perun-base/src/main/java/cz/metacentrum/perun/core/api/AttributeDefinition.java]: It probably doesn't have to be part of this pull-request, but CLI a GUI must reflect new AttributeDefinition property. Please add it (also to print and create tools) or make a note and this can be merged only after change is done in separate pull-request.

File: perun-base/src/test/resources/test-schema.sql [hunk 7]
```diff
@@ -180,17 +180,18 @@ create table groups (
 );
 
 create table members (
-	id integer not null,
-	user_id integer not null,
-	vo_id integer not null,
+	id integer not null PRIMARY KEY,
```
Comment (zlamalp) [inline perun-base/src/test/resources/test-schema.sql]: As @stavamichal stated, we define constraints (with exception of not null and default) after table definition, so we can name them. I believe, it has something to do with naming automatically created indexes etc. Since we change schema from time to time, we need to now, ho to explicitly address anything.

File: perun-base/src/test/resources/test-schema.sql [hunk 8]
```diff
@@ -180,17 +180,18 @@ create table groups (
...
+	user_id integer not null REFERENCES users(id) ON DELETE CASCADE,
+	vo_id integer not null REFERENCES vos(id) ON DELETE CASCADE,
```
Comment (zlamalp) [inline perun-base/src/test/resources/test-schema.sql]: We do not use ``ON DELETE CASCADE``, since we want to audit all delete actions and remove all relations in Bl. I believe, it's used only in registrar, for the simplification of actions in core (since core can't reach registrar methods and logic would have to be duplicated).

File: perun-base/src/test/resources/test-schema.sql [hunk 9]
```diff
@@ -1685,3 +1673,120 @@ alter table user_ext_source_attr_values add constraint uesattrval_attr_fk foreig
 
 alter table members_sponsored add constraint memspons_mem_fk foreign key (sponsored_id) references members(id);
 alter table members_sponsored add constraint memspons_usr_fk foreign key (sponsor_id) references users(id);
+
+-- these tables must be created after all the referenced primary keys are defined
```
Comment (zlamalp) [inline perun-base/src/test/resources/test-schema.sql]: I believe you should be able to create tables and add constraints and references after they are added to the related tables. So keep it at the end of create table section and add constraints at the end of alter table section.

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java [hunk 10]
```diff
@@ -3621,14 +3621,14 @@
...
 
 	/**
 	 * PRIVILEGE: Remove attributes only when principal has access to write on them.
 	 *
 	 * Batch version of removeAttribute. This method automatically skip all core attributes which can't be removed this way.
 	 * @throws AttributeNotExistsException if the any of attributes doesn't exists in underlying data source
-	 * @see cz.metacentrum.perun.core.api.AttributesManager#removeAttribute(PerunSession, Resource, Member, AttributeDefinition)
+	 * @see AttributesManager#removeAttribute(PerunSession, Member, Resource, AttributeDefinition)
 	 */
 	void removeAttributes(PerunSession sess, Resource resource, Member member, List<? extends AttributeDefinition> attributes) throws InternalErrorException, PrivilegeException, AttributeNotExistsException, MemberNotExistsException, ResourceNotExistsException, WrongAttributeAssignmentException, WrongAttributeValueException, WrongReferenceAttributeValueException, MemberResourceMismatchException;
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java]: I think, there are other methods for manipulating attributes too, like checkAttribute(s)/fillAttribute(s), so I suggest to correct order in all API for user-facility, member-resource, member-group or whatever attributes. Or just scratch it and we will correct API in own pull-request, since it might very easily conflict with work of others.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java [hunk 11]
```diff
@@ -3982,5 +3985,42 @@ public void mergeAttributesValues(PerunSession sess, Member member, List<Attribu
...
+	/**
+	 * Finds pair of ids of Perunbeans that has the attribute's value for the attribute.
+	 * The same as {@link #getPerunBeanIdForUniqueAttributeValue(PerunSession, Attribute)} for attributes assigned to pairs of beans.
+	 * E.g. for attribute with group_resource namespace, groupId and resourceId would be returned.
+	 *
+	 * @param sess session
+	 * @param attribute attribute with a filled value that will be checked for uniqueness
+	 * @return pair of PerunBean ids (order depends on attribute namespace, e.g. for group_resource it would be groupId,resourceId) or null if not found
+	 */
+	Pair<Integer,Integer> getTwoPerunBeanIdsForUniqueAttributeValue(PerunSession sess, Attribute attribute) throws InternalErrorException;
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java]: I believe we have some check methods for attribute namespaces in Impl or some utils, maybe we could have this logic together somewhere.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/Auditer.java [hunk 12]
```diff
@@ -171,13 +169,7 @@ public Auditer() {
...
 	public void setPerunPool(DataSource perunPool) throws InternalErrorException {
 		this.jdbc = new JdbcPerunTemplate(perunPool);
-		if(Compatibility.isOracle()) {
-			OracleLobHandler oracleLobHandler = new OracleLobHandler();
-			oracleLobHandler.setNativeJdbcExtractor(new CommonsDbcpNativeJdbcExtractor());
-			lobHandler = oracleLobHandler;
-		} else {
-			lobHandler = new DefaultLobHandler();
-		}
+		this.lobHandler = new DefaultLobHandler();
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/Auditer.java]: I'm not even sure, we need LobHandler at all, since we only store (N)CLOB (for Oracle) and it can be read as ``.getString()`` by JDBC. But it probably shouldn't be part of this change. Just suggestions to test and implement.

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java]: OK, I would have to change the order at 50 places just for group-resource, so I am going to scratch the change.

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/Attribute.java (hunk 5 above)
Comment (martin-kuba) [reply perun-base/src/main/java/cz/metacentrum/perun/core/api/Attribute.java]: No, these methods are not properties because their names do not start with "get".

File: perun-base/src/test/resources/test-schema.sql (hunk 8 above)
Comment (martin-kuba) [reply perun-base/src/test/resources/test-schema.sql]: OK, removed

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java [hunk 13]
```diff
@@ -4158,4 +4158,15 @@
 	 * @throws AttributeNotExistsException when attribute IDs in rights don't refer to existing attributes
 	 */
 	void setAttributeRights(PerunSession sess, List<AttributeRights> rights) throws InternalErrorException, PrivilegeException, AttributeNotExistsException;
+
+	/**
+	 * Converts attribute to unique.
+	 * Marks the attribute definitiona as unique, and copies all values to a special table with unique constraint
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java]: fixed

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java [hunk 14]
```diff
@@ -4158,4 +4158,15 @@
...
+	 * that ensures that all values remain unique. Values of type ArrayList and LinkedHashMap are splitted into
+	 * multiple values, thus each of the subvalues must be unique.
+	 * @param session perun session
+	 * @param attrId attribute id
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/AttributesManager.java]: I do think that listing exceptions without their cause (as in previous methods) just pollutes the javadoc. The causes are clear from the names of exceptions.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java [hunk 15]
```diff
@@ -1497,7 +1500,7 @@
 	 * @throws WrongReferenceAttributeValueException
 	 * @throws MemberResourceMismatchException
 	 */
-	void setAttribute(PerunSession sess, Resource resource, Member member, Attribute attribute) throws InternalErrorException, WrongAttributeValueException, WrongAttributeAssignmentException, WrongReferenceAttributeValueException, MemberResourceMismatchException;
+	void setAttribute(PerunSession sess, Member member, Resource resource, Attribute attribute) throws InternalErrorException, WrongAttributeValueException, WrongAttributeAssignmentException, WrongReferenceAttributeValueException, MemberResourceMismatchException;
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java]: I have scratched the change.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java]: OK, I have collapsed the two methods into one.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 16]
```diff
@@ -6398,6 +6399,34 @@ public void checkGroupIsFromTheSameVoLikeResource(PerunSession sess, Group group
 			throw new GroupResourceMismatchException("Group is not from the same vo like Resource: " + group + " " + resource);
 	}
 
+	@Override
+	public Integer getPerunBeanIdForUniqueAttributeValue(PerunSession sess, Attribute attribute) throws InternalErrorException {
+		if(!attribute.isUnique()) {
+			throw new InternalErrorException("attribute definition is not marked as unique: "+attribute);
+		}
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: removed the method

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 17]
```diff
@@ -6398,6 +6399,34 @@ public void checkGroupIsFromTheSameVoLikeResource(PerunSession sess, Group group
...
+		return attributesManagerImpl.getPerunBeanIdForUniqueAttributeValue(sess, attribute);
+	}
+
+	@Override
+	public Pair<Integer, Integer> getTwoPerunBeanIdsForUniqueAttributeValue(PerunSession sess, Attribute attribute) throws InternalErrorException {
+		if(!attribute.isUnique()) {
+			throw new InternalErrorException("attribute definition is not marked as unique: "+attribute);
+		}
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: removed the method

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/Auditer.java (hunk 12 above)
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/Auditer.java]: I do not know, I have just removed the branch for Oracle because ti was complaining that default lob handler should be used since some version of Oracle JDBC driver.

File: perun-db/oracle.sql (hunk 3 above)
Comment (martin-kuba) [reply perun-db/oracle.sql]: OK, I am going to change it whole, just do not complain that the change is too big :-)

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 18]
```diff
@@ -4048,13 +4030,208 @@ public void registerVirtAttributeModules(ServiceLoader<AttributesModuleImplApi>
...
+								for (Map.Entry<String, String> entry : map.entrySet()) {
+									jdbc.update("INSERT INTO " + tablePrefix + "_attr_u_values (" + bean + "_id,attr_id,attr_value) VALUES (?,?,?)", beanId, attrDef.getId(), entry.getKey() + "=" + entry.getValue());
+								}
+								break;
+						}
+						int c = counter.addAndGet(1);
+						if(c%1000==0) log.info(""+c+" values of "+attrDef.getName()+" were converted");
+
+					} catch (InternalErrorException e) {
+						log.error("problem converting attribute to unique", e);
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: it should be re-thrown as a runtime exception

File: perun-base/src/main/java/cz/metacentrum/perun/core/api/AttributeDefinition.java (hunk 6 above)
Comment (martin-kuba) [reply perun-base/src/main/java/cz/metacentrum/perun/core/api/AttributeDefinition.java]: I have created issue for this - https://homeproj.cesnet.cz/issues/3948

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java [hunk 19]
```diff
@@ -3982,5 +3985,42 @@ public void mergeAttributesValues(PerunSession sess, Member member, List<Attribu
...
+
+	/**
+	 * Finds id of a PerunBean that has the attribute's value for the attribute. The attribute must be marked as unique.
+	 * For simple value types (String,Integer,Boolean), id of PerunBean that already has this value assigned is returned, or null,
+	 * if not found. For composite values types (ArrayList, LinkedHashMap), the first PerunBean that has any of
+	 * the simple values contained in the composite value is returned, or null if not found.
+	 *
+	 * This method is intended for finding whether all of the values in a list or map are unique among all PerunBeans
+	 * for a given attribute definition. If an id is returned, some of the simple values is already assigned
+	 * to a PerunBean.
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AttributesManagerBl.java]: I have added more detailed explanation for maps.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 20]
```diff
@@ -6398,6 +6399,34 @@ public void checkGroupIsFromTheSameVoLikeResource(PerunSession sess, Group group
...
+		return attributesManagerImpl.getTwoPerunBeanIdsForUniqueAttributeValue(sess, attribute);
+	}
+
+	@Override
+	public void convertAttributeToUnique(PerunSession session, int attrId) throws InternalErrorException, AttributeNotExistsException, AttributeAlreadyMarkedUniqueException {
+		AttributeDefinition attrDef = getAttributeDefinitionById(session, attrId);
+		if(attrDef.isUnique()) throw new AttributeAlreadyMarkedUniqueException("cannot convert attribute", attrDef);
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: OK

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 21]
```diff
@@ -1737,17 +1684,20 @@ public boolean setAttribute(final PerunSession sess, final Object object, final
...
-		params.put("attr_id", attribute.getId());
-		params.put(columnName, identificator);
+		// create lists of parameters for the where clause of the SQL query
+		List<String> columnNames = Arrays.asList( "attr_id", columnName);
+		List<Object> columnValues = Arrays.asList( attribute.getId(), identificator);
 
 		// save attribute
-		return setAttributeInDB(sess, attribute, tableName, params);
+		boolean changedDb = setAttributeInDB(sess, attribute, tableName, columnNames, columnValues);
+		if(changedDb && attribute.isUnique() && (object instanceof PerunBean)) {
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: added

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 22]
```diff
@@ -2002,19 +1989,14 @@ public AttributeDefinition createAttribute(PerunSession sess, AttributeDefinitio
 
 	}
 
+	private static String attr2tablePrefix(AttributeDefinition attribute) {
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: OK

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 23]
```diff
@@ -2002,19 +1989,14 @@ public AttributeDefinition createAttribute(PerunSession sess, AttributeDefinitio
...
+		return ENTITIES_TO_BEANS_MAP.get(attribute.getNamespace().split(":")[2]);
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: It will work, the map converts it.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 24]
```diff
@@ -4048,13 +4030,208 @@ public void registerVirtAttributeModules(ServiceLoader<AttributesModuleImplApi>
...
+		try {
+			jdbc.update("UPDATE attr_names SET is_unique="+Compatibility.getTrue()+", modified_by=?, modified_by_uid=?, modified_at="
+					+ Compatibility.getSysdate() + " WHERE id=?",session.getPerunPrincipal().getActor(), session.getPerunPrincipal().getUserId(),attrDef.getId());
+		} catch (RuntimeException ex) {
+			throw new InternalErrorException(ex);
+		}
+	}
+
+	@Override
+	public void convertAttributeValuesToUnique(PerunSession session, AttributeDefinition attrDef) throws InternalErrorException {
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: No, it throws exception.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (stavamichal) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: Great. Just please, add it also to API Javadoc and RPC Javadoc (if it is also missing). Thank you.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (stavamichal) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: You are right, I missed that.

**Comment by martin-kuba** [comment]
I think all the changes requested in reviews were done.

**Comment by martin-kuba** [comment]
Added AttributeDefinition .unique to audit messages.
